// stdout-stderr/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const stripAnsi = require("strip-ansi");
const debug = require('debug')('stdout-stderr');
const g = global;
if (!g['stdout-stderr']) {
    g['stdout-stderr'] = {
        stdout: process.stdout.write,
        stderr: process.stderr.write,
    };
}
/** mocks stdout or stderr */
function mock(std) {
    let writes = [];
    return {
        stripColor: true,
        print: false,
        start() {
            debug('start', std);
            writes = [];
            process[std].write = (data, ...args) => {
                writes.push(bufToString(data));
                if (this.print)
                    g['stdout-stderr'][std].apply(process[std], [data, ...args]);
                return true;
            };
        },
        stop() {
            process[std].write = g['stdout-stderr'][std];
            debug('stop', std);
        },
        get output() {
            let o = this.stripColor ? writes.map(stripAnsi) : writes;
            return o.join('');
        },
    };
}
exports.stdout = mock('stdout');
exports.stderr = mock('stderr');
function bufToString(b) {
    if (typeof b === 'string')
        return b;
    return b.toString('utf8');
}
